packages/web3-eth-contract/src/index.js

```diff
@@ -581,9 +581,12 @@ Contract.prototype.deploy = function(options, callback){
     options = this._getOrSetDefaultOptions(options);
 
 
-    // return error, if no "data" is specified
+    // throw error, if no "data" is specified
     if(!options.data) {
-        return utils._fireError(new Error('No "data" specified in neither the given options, nor the default options.'), null, null, callback);
+        if (typeof callback === 'function'){
+            return callback(errors.ContractMissingDeployDataError());
+        }
+        throw errors.ContractMissingDeployDataError();
```

Comment:
This block is the only real logic change. Fixes bug where `utils._fireError` is usually swallowing this error.

`contract.deploy` doesn't have an event emitter or a promise associated with it (and AFAIK, isn't really invoked with a callback) 

Have rewritten it into a thrown error but kept backwards compat for the callback in case anyone actually doing that.

Reply:
Isn't it possible to handle this the same way as on all other places e.g.:
```
utils._fireError(errors.ContractMissingFromAddressError(), defer.eventEmitter, defer.reject, args.callback);
```

Reply:
This just because of consistency reasons. In the future, it can have some advantages for us.

Reply:
I'm not sure, actually. `deploy` isn't async - it returns a transaction object. The eventEmitter and promise args to _fireError are explicitly set to null in the original.

So this is a pre-flight error that's accidentally getting swallowed by utils._fireError when it's invoked synchronously like
```
myContract.deploy(options);
myContract.deploy(options).encodeABI(); 
```